cmd/syncthing/gui.go

```diff
@@ -367,12 +375,20 @@ func (s *apiService) Serve() {
 		close(s.startedOnce)
 	}
 
-	// Serve in the background
+	// Serve in the background for each listener
 
 	serveError := make(chan error, 1)
-	go func() {
-		serveError <- srv.Serve(listener)
-	}()
+	srv := http.Server{
+		Handler: handler,
+		// ReadTimeout must be longer than SyncthingController $scope.refresh
+		// interval to avoid HTTP keepalive/GUI refresh race.
+		ReadTimeout: 15 * time.Second,
+	}
+	for _, listener := range listeners {
+		go func(listener net.Listener) {
+			serveError <- srv.Serve(listener)
+		}(listener)
+	}
```

Comment:
This will leave zombie routines hanging on channel send for every other listener apart from the first one that fails.

Reply:
Go concurrency paradigm is still an area where I do not yet feel comfortable. ;-)
How would you better handle all these goroutine lifetimes?

Reply:
So there are a few approaches, depending on your appetite to learn.

Easiest is to do:
```
err := srv.Serve(listener)
select {
  serverError <- err:
  default:
}
```

So that we discard errors once the channel is full.
As the loop exits, the channel would get GCed.

More idiomatic way would be to have a wait group, add 1 for every listener, create error channel with a buffer of `len(listeners)`. Upon receiving the first error (or the stop), close all listeners, wait for the waitgroup to resolve (all listeners to exit), drain the error channel (discarding errors), close the channel, and return.